cumulus/pallets/parachain-system/src/block_weight/transaction_extension.rs

```diff
@@ -0,0 +1,489 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Cumulus.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::{
+	block_weight_over_target_block_weight, is_first_block_in_core_with_digest, BlockWeightMode,
+	MaxParachainBlockWeight, LOG_TARGET,
+};
+use crate::WeightInfo;
+use alloc::vec::Vec;
+use codec::{Decode, DecodeWithMemTracking, Encode};
+use cumulus_primitives_core::CumulusDigestItem;
+use frame_support::{
+	dispatch::{DispatchClass, DispatchInfo, PostDispatchInfo},
+	pallet_prelude::{
+		InvalidTransaction, TransactionSource, TransactionValidityError, ValidTransaction,
+	},
+	weights::Weight,
+};
+use scale_info::TypeInfo;
+use sp_core::Get;
+use sp_runtime::{
+	traits::{DispatchInfoOf, Dispatchable, Implication, PostDispatchInfoOf, TransactionExtension},
+	DispatchResult,
+};
+
+/// Transaction extension that dynamically changes the max block weight.
+///
+/// With block bundling, parachains are running with block weights that may not allow certain
+/// transactions to be applied, e.g. a runtime upgrade. To ensure that these transactions can still
+/// be applied, this transaction extension can change the max block weight as required. There are
+/// multiple requirements for it to change the block weight:
+///
+/// 1. Only the first block of a core is allowed to change its block weight.
+///
+/// 2. Any `inherent` or any transaction up to `MAX_TRANSACTION_TO_CONSIDER` requires more block
+///    weight than the target block weight. Target block weight is the max weight for the respective
+///    extrinsic class.
+///
+/// Because the node is tracking the wall clock time while building a block to abort block
+/// production if it takes too long, we do not allow any block to change the block weight. The node
+/// knows that the first block of a core may runs longer. So, the node allows this block to take up
+/// to `2s` of wall clock time. `2s` is the time each `PoV` gets on the relay chain for its
+/// validation or in other words the maximum core execution time. The extension sets the
+/// [`CumulusDigestItem::UseFullCore`] digest when the block should occupy the entire core.
```

Comment:
I just realized that you did not modify any authoring logic. In theory, the authorship code could query some runtime API or on_init return value to check how long the block could take at most and adjust the timeout.